ceph/rados/mgr_workflows.py

```diff
@@ -0,0 +1,274 @@
+"""
+Module to perform Serviceability scenarios on mgr daemons
+"""
+
+import datetime
+import time
+
+from ceph.ceph_admin import CephAdmin
+from ceph.rados.core_workflows import RadosOrchestrator
+from ceph.rados.monitor_workflows import MonitorWorkflows
+from utility.log import Log
+
+log = Log(__name__)
+
+
+class MgrWorkflows:
+    """
+    Module to perform Serviceability scenarios on mgr daemons
+    """
+
+    def __init__(self, node: CephAdmin):
+        """
+        Module to perform Serviceability scenarios on mgr daemons
+
+        Args:
+            node: CephAdmin object
+        """
+        self.rados_obj = RadosOrchestrator(node=node)
+        self.mon_obj = MonitorWorkflows(node=node)
+        self.cluster = node.cluster
+        self.config = node.config
+        self.client = node.cluster.get_nodes(role="client")[0]
+        self.cluster_nodes = node.cluster.get_nodes()
+
+    def add_mgr_service(self, **kwargs) -> bool:
+        """
+        Method to add the mgr service from the given host
+        Args:
+            kwargs: the accepted KW args for the method are:
+                host: Cluster object of the host from where mgr needs to be Added
+        returns:
+            Pass -> True, Fail -> false
+        """
+        # Checking if mgr service exists before removing
+        host = kwargs.get("host")
+        hostname = host.hostname
+
+        if self.check_mgr_exists_on_host(host=hostname):
+            log.info(
+                f"MGR daemon already present on the host {hostname}. Returning Pass"
+            )
+            return True
+
+        log.debug(f"Adding mgr daemon on the host : {hostname}")
+
+        # Adding lable to host
+        labels = self.mon_obj.get_host_labels(host=hostname)
+        if "mgr" not in labels:
+            cmd = f"ceph orch host label add {hostname} mgr"
+            self.client.exec_command(sudo=True, cmd=cmd)
+            time.sleep(10)
+            log.debug(f"Added mgr label for the host : {hostname}")
+
+        # Adding host as mgr
+        cmd = "ceph orch apply mgr --placement='label:mgr'"
```

Comment:
As long as the service is managed by cepadm, and deployment is based on labels, we should not have to add the service again with ` ceph orch apply ` command. 

We can add a check to see if the mgr daemon is automatically added in X duration after adding the mgr label on the host.

Reply:
This step is required, or else the addition of MGR will fail. I executed the code by commenting on the command, and the test execution failed.
Log- http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-WNJFBP

Reply:
I did test it, and if the service is managed by cephadm, the daemon deployment should be done automatically.

For now, this should be fine. We can check the code later.

Reply:
Here is the root cause of why it's failed-
1.  From the ceph status  we can notice that there are two mgrs
    `  mgr: ceph-rados-auto-mtptcq-node1-installer.fbbadx(active, since 2h), standbys: ceph-rados-auto-mtptcq-node6.pbijfa`

2.  From the "ceph orch ls --export"  we can observe that the  mgr count in service spec  is assigned to 2
    
```
service_name: mgr
placement:
  count: 2
  hosts:
  - ceph-rados-auto-mtptcq-node1-installer
  - ceph-rados-auto-mtptcq-node6
```

3. Adding the mgr label to the node.  At this step, the mgr is not configured because the count in the service spec is 2
  `  ceph orch host label add  mgr` . Now there are 4 nodes with the mgr label.

    services in ceph status:
         ```
services:
    mon: 3 daemons, quorum ceph-rados-auto-mtptcq-node1-installer,ceph-rados-auto-mtptcq-node6,ceph-rados-auto-mtptcq-node2 (age 3h)
    mgr: ceph-rados-auto-mtptcq-node1-installer.fbbadx(active, since 2h), standbys: ceph-rados-auto-mtptcq-node6.pbijfa
    mds: 1/1 daemons up, 1 standby
```

[ceph: root@ceph-rados-auto-mtptcq-node1-installer /]# ceph orch host ls
HOST                                    ADDR          LABELS                                                                        STATUS  
ceph-rados-auto-mtptcq-node1-installer  10.0.209.235  _admin,installer,node-exporter,grafana,mgr,mon,alertmanager,crash,prometheus          
ceph-rados-auto-mtptcq-node2            10.0.210.33   node-exporter,rgw,mds,mgr,mon,alertmanager,crash,nfs                                  
ceph-rados-auto-mtptcq-node3            10.0.210.159  crash,osd,node-exporter,mgr                                                           
ceph-rados-auto-mtptcq-node4            10.0.211.177  crash,osd,node-exporter                                                               
ceph-rados-auto-mtptcq-node5            10.0.208.33   crash,osd,node-exporter                                                               
ceph-rados-auto-mtptcq-node6            10.0.209.0    node-exporter,rgw,mds,mgr,mon,crash,nfs                                               
6 hosts in cluster

4.  After executing the  "ceph orch apply mgr --placement='label:mgr'" command, all four nodes which are label mgr are configured with the mgr daemons and services spec also updated accordingly.

services in the ceph status:

services:
    mon: 3 daemons, quorum ceph-rados-auto-mtptcq-node1-installer,ceph-rados-auto-mtptcq-node6,ceph-rados-auto-mtptcq-node2 (age 3h)
    mgr: ceph-rados-auto-mtptcq-node1-installer.fbbadx(active, since 2h), standbys: ceph-rados-auto-mtptcq-node6.pbijfa, ceph-rados-auto-mtptcq-node3.gmbggw, ceph-rados-auto-mtptcq-node2.rjuljm
    mds: 1/1 daemons up, 1 standby
 
service spec:

```
 ---
service_type: mgr
service_name: mgr
placement:
  label: mgr
```
5.  Now by adding the label  to host will configure the mgr daemon.